storage/imxrt-flash/flashsrv.c

```diff
@@ -94,100 +306,363 @@ static void flashsrv_devCtl(msg_t *msg)
+					LOG_ERROR("imxrt-flashsrv: request goes beyond parition range, pSize: %d, write offs: %d.", memory->parts[msg.i.io.oid.id].sz, msg.i.io.offs);
+					break;
+				}
+
+				mutexLock(memory->lock);
+				msg.o.io.err = flashsrv_bufferedPagesWrite(memory->fOid.id, beginAddr + msg.i.io.offs, msg.i.data, msg.i.size);
+				mutexUnlock(memory->lock);
+				break;
+
+			case mtDevCtl:
+				mutexLock(memory->lock);
+				flashsrv_rawCtl(memory, &msg);
+				mutexUnlock(memory->lock);
+				break;
+
+			default:
+				msg.o.io.err = -ENOSYS;
+				break;
+		}
+		msgRespond(memory->rawPort, &msg, rid);
+	}
 }
 
 
 static void flashsrv_devThread(void *arg)
 {
 	msg_t msg;
 	unsigned rid;
-	uint32_t port = (uint32_t)arg;
+
+	flash_memory_t *memory = (flash_memory_t *)arg;
 
 	while (1) {
-		while (msgRecv(port, &msg, &rid) < 0)
+		while (msgRecv(memory->fOid.port, &msg, &rid) < 0)
 			;
 
 		switch (msg.type) {
 			case mtRead:
-				msg.o.io.err = flashsrv_read(msg.i.io.oid, msg.i.io.offs, msg.o.data, msg.o.size);
+				mutexLock(memory->lock);
+				msg.o.io.err = flashsrv_read(memory->fOid.id, msg.i.io.offs, msg.o.data, msg.o.size);
+				mutexUnlock(memory->lock);
 				break;
 
 			case mtWrite:
-				msg.o.io.err = flashsrv_write(msg.i.io.oid, msg.i.io.offs, msg.i.data, msg.i.size);
+				mutexLock(memory->lock);
+				msg.o.io.err = flashsrv_bufferedPagesWrite(memory->fOid.id, msg.i.io.offs, msg.i.data, msg.i.size);
+				mutexUnlock(memory->lock);
 				break;
 
 			case mtDevCtl:
-				flashsrv_devCtl(&msg);
+				mutexLock(memory->lock);
+				flashsrv_devCtl(memory, &msg);
+				mutexUnlock(memory->lock);
 				break;
 
 			default:
 				msg.o.io.err = -ENOSYS;
 				break;
 		}
-		msgRespond(port, &msg, rid);
+		msgRespond(memory->fOid.port, &msg, rid);
+	}
+}
+
+
+/* Initialization functions */
+
+static int flashsrv_initMeterfs(flashsrv_partition_t *part)
+{
+	meterfs_ctx_t *ctx;
+
+	if ((part->fsCtx = calloc(1, sizeof(meterfs_ctx_t))) == NULL) {
+		LOG_ERROR("imxrt-flashsrv: cannot allocate memory.");
+		return -ENOMEM;
+	}
+
+	ctx = (meterfs_ctx_t *)part->fsCtx;
+
+	ctx->sz = part->sz;
+	ctx->offset = part->offs;
+	ctx->sectorsz = flashsrv_common.flash_memories[part->fID].ctx.properties.sector_size;
+
+	if (part->fID == 0) {
+		ctx->read = flashsrv_mfsReadf0;
+		ctx->write = flashsrv_mfsWritef0;
+		ctx->partitionErase = flashsrv_mfsPartitionErasef0;
+		ctx->eraseSector = flashsrv_mfsEraseSectorf0;
+	}
+	else if (part->fID == 1) {
+		ctx->read = flashsrv_mfsReadf1;
+		ctx->write = flashsrv_mfsWritef1;
+		ctx->partitionErase = flashsrv_mfsPartitionErasef1;
+		ctx->eraseSector = flashsrv_mfsEraseSectorf1;
+	}
+
+	if (meterfs_init(ctx) < 0) {
+		LOG_ERROR("imxrt-flashsrv: init meterfs at flash: %d, partition: %d.", part->fID, part->oid.id);
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int flashsrv_mountPart(ptable_partition_t *pHeader, flashsrv_partition_t *part)
+{
+	int res = EOK;
+	char path[32];
+
+	sprintf(path, "/dev/flash%d.%s", part->fID, pHeader->name);
```

Comment:
Potential stack corruption (path overflow). sprintf(path, "/dev/flash%d.%19s perhaps?

Reply:
- use `%u` for printing unsigned values (this applies in other places across the code)
- if `pHeader->name` won't contain `\0` it will corrupt the stack (the partition names are not checked to contain only alphanumeric chars either)
- I would personally use `snprintf` as this is the preferred way of ensuring we won't harm ourselfs